compiler/src/openxla/compiler/nvgpu/Dialect/TritonFlow/Transforms/OutlineTritonCalls.cpp

```diff
@@ -28,12 +38,135 @@ using namespace mlir;
 
 namespace openxla::compiler::nvgpu::tritonflow {
 
+// TODO(ezhulenev): Add support for tied results.
+
+// Returns the IREE HAL pipeline layout inferred from the Triton call operation.
+static IREE::HAL::PipelineLayoutAttr getPipelineLayout(tritonflow::CallOp call,
+                                                       triton::FuncOp callee) {
+  MLIRContext* ctx = callee->getContext();
+
+  // Scalar arguments passed as dispatch constants.
+  int64_t pushConstants = llvm::count_if(
+      callee.getArgumentTypes(), [](Type type) { return type.isIntOrIndex(); });
+
+  // Bindings for all tensor (buffer) arguments.
+  llvm::SmallVector<IREE::HAL::DescriptorSetBindingAttr> bindings;
+
+  // Add tensor (buffer) arguments as ReadOnly bindings.
+  for (auto arg : call.getArguments().getTypes()) {
+    if (auto tensor = arg.dyn_cast<RankedTensorType>()) {
+      bindings.push_back(IREE::HAL::DescriptorSetBindingAttr::get(
+          ctx, /*ordinal=*/bindings.size(),
+          IREE::HAL::DescriptorType::StorageBuffer,
+          IREE::HAL::DescriptorFlags::ReadOnly));
+    }
+  }
+
+  // Add results as bindings with default flags.
+  for (auto ret : call.getResultTypes()) {
+    if (auto tensor = ret.dyn_cast<RankedTensorType>()) {
+      bindings.push_back(IREE::HAL::DescriptorSetBindingAttr::get(
+          ctx, /*ordinal=*/bindings.size(),
+          IREE::HAL::DescriptorType::StorageBuffer, std::nullopt));
+    }
+  }
+
+  return IREE::HAL::PipelineLayoutAttr::get(
+      ctx, pushConstants,
+      IREE::HAL::DescriptorSetLayoutAttr::get(ctx, /*ordinal=*/0, bindings));
+}
+
+// Updates Triton function signature to be compatible with IREE dispatch ABI. If
+// function type was updated, returns an argument permutation that was applied.
+static std::optional<SmallVector<int64_t>> updateTritonFunctionForAbi(
+    triton::FuncOp callee) {
+  MLIRContext* ctx = callee->getContext();
+
+  // We need to compute permutation of function arguments to update callsites.
+  struct IndexedArg {
+    size_t index;
+    Type type;
+  };
+
+  SmallVector<IndexedArg> args;
+  for (auto indexed : llvm::enumerate(callee.getArgumentTypes()))
+    args.push_back({indexed.index(), indexed.value()});
+
+  // Partition arguments types, so that all scalars pushed to the end.
+  auto firstScalarArg = std::stable_partition(
+      args.begin(), args.end(),
+      [](IndexedArg indexed) { return !indexed.type.isIntOrIndex(); });
```

Comment:
Would it make sense to pull `type.getIntOrIndex()` into a `isScalarArg()` function (I've seen one other occurrence above)?

Reply:
Done